File: R/tagtools/R/crop_to.R
#' Reduce the time span of data
#' 
#' This function is used to reduce the time span of data by cropping out any data that falls before and after two time cues.
#' 
#' @param X A sensor list, vector, or matrix. X can be regularly or irregularly samples data in any frame and unit.
#' @param sampling_rate The sampling rate of X in Hz. This is only needed if X is not a sensor structure.
#' @param T A vector of sampling times for X. This is only needed if X is not a sensor list and X is not regularly sampled. 
#' @param tcues A two-element vector containing the start and end time cues in seconds of the data segment to keep (i.e., tcues <- c(start_time, end_time)).
#' @return Cropped data in the same format as X, unless X is irregularly sampled and NOT a sensor list. In that case, the function returns a list with 2 elements:
#' \itemize{
#'  \item{\strong{X: }} A sensor list, vector or matrix containing the cropped data segment. If the input is a sensor list, the output will also be. The output has the same units, frame and sampling characteristics as the input.
#'  \item{\strong{T: }} A vector of sampling times for Y. This is only returned if X is irregularly sampled and X is not a sensor list. (If X is a sensor list, the sampling times are stored in the list.)
#' }
#' @examples 
#'          d <- find_dives(beaked_whale$P,300) 
#'          P2 <- crop_to(beaked_whale$P, tcues = c(d$start[1], d$end[1]))	#crop to 1st dive
#'          plott(list(P2$X), r=c(1), panel_labels=c('Depth'))
#'          #plot shows the dive profile of the selected dive
#' @export

crop_to <- function(X, sampling_rate = NULL, tcues, T = NULL) {
  T <- c()
  if (is.list(X)) {
    if (missing(tcues)) {
      stop("input for tcues is required for crop_to\n")
    }
    x <- X$data
    sampling_rate <- X$sampling_rate
    if (!is.matrix(x)) {
      x <- matrix(x, ncol = 1)
    }
  } else {
    if (missing(tcues)) {
      stop("inputs for X, sampling_rate, and tcues are all required")
    }
    x <- X
    if (!is.matrix(x)) {
      x <- matrix(x, ncol = 1)
    }
    if (nrow(x) == 1) {
      x <- t(x)
    }
  }
  if (length(tcues) != 2) {
    stop("tcues must be a two-element vector of c(start_time,end_time)")
  }
  tcues <- sort(tcues)
  
  if (!is.null(T)) {    #irregularly sampled data
    k <- which(T >= tcues[1] & T <= tcues[2])
    T <- T[k] - tcues[1]
  } else {
    k <- c(max(round(tcues[1] * sampling_rate), 1):min(round(tcues[2] * sampling_rate), nrow(x)))
  }
  
  if (length(k) == 0) {
    stop('No data points observed between requested time cues.\n')
  }
  if ((k[1] <= 1) & (k[length(k)] >= nrow(x))) {
    if (!is.null(T)){
      return(list(X = X, T = T))
    }else{
      return(X)
    }
  }
  
  if (!is.list(X)) {
    X <- x[k, ]
    if (!is.null(T)){
      return(list(X = X, T = T))
    }else{
      return(X)
    }
  }
  
  if (length(T) > 1) {
    X$data <- cbind(T, x[k, ])
  } else {
    X$data <- x[k, ]
  }
  
  X$crop <- tcues
  X$crop_units <- "seconds"
  X$start_time <- tcues[1]
  X$start_time_units <- "seconds"
  if (("history" %in% names(X) == FALSE) | is.null(X$history)) {
    X$history <- "crop_to"
  } else {
    X$history <- paste(X$history, "crop_to", sep = ',')
  }
  
  return(X)
}

File: matlab/calibration/crop_to.m
function		[X,T] = crop_to(X,fs,tcues)

%		Y = crop_to(X,tcues)				% X is a sensor structure
%		or
%		Y = crop_to(X,fs,tcues)			% X is a regularly sampled vector or matrix
%		or
%		[Y,T] = crop_to(X,T,tcues)		% X is an irregularly sampled vector or matrix
%
%		Reduce the time span of data by cropping out any data that falls before and
%		after two time cues.
%
%     Inputs:
%     X is a sensor structure, vector or matrix. X can be regularly or
%		 irregularly sampled data in any frame and unit.
%     fs is the sampling rate of X in Hz. This is only needed if
%		 X is not a sensor structure and X is regularly sampled.
%		T is a vector of sampling times for X. This is only needed if X is
%		 not a sensor structure and X is not regularly sampled.
%		tcues is a two-element vector containing the start and end time cue
%		 in seconds of the data segment to keep, i.e., tcues = [start_time, end_time].
%
%     Results:
%     Y is a sensor structure, vector or matrix containing the cropped data segment.
%		 If the input is a sensor structure, the output will also be. The output has
%		 the same units, frame and sampling characteristics as the input.
%     T is a vector of sampling times for Y. This is only returned if X is irregularly
%	    sampled and X is not a sensor structure. If X is a sensor structure, the sampling
%		 times are stored in the structure.
%
%		Example:
%		 load_nc('testset3');
%		 d = find_dives(P,300) ;
%		 P2 = crop_to(P,[d.start(2) d.end(2)]);	% crop to 2nd dive
%		 plott(P2)
%		 % plot shows the profile of the second dive
%
%     Valid: Matlab, Octave
%     markjohnson@st-andrews.ac.uk
%     Last modified: 23 Dec 2018
%     - changed sensor structure field names

T = [] ;
if nargin<2,
	help crop_to
	return
end
	
if isstruct(X),
	[x,fst] = sens2var(X) ;
	if isempty(x), return, end
	tcues = fs ;
	fs = fst ;
else
	if nargin<3,
		help crop_to
		return
	end
	x = X ;
	if size(x,1)==1,		% make sure x is a column vector
		x = x(:) ;
	end
end

if length(tcues)~=2,
	fprintf(' crop_to: tcues must be a two-element vector of [start_time,end_time]\n') ;
	return
end

if tcues(1)>=tcues(2),
	X = [] ;
	return
end
	
if length(fs)>1,		% irregularly sampled data
	k = find(fs>=tcues(1) & fs<=tcues(2)) ;
	T = fs(k)-tcues(1) ;
else
	k = max(round(tcues(1)*fs)+1,1):min(round(tcues(2)*fs)+1,size(x,1)) ;
end

if isempty(k),								% crop excludes all of the input data
	X = [] ; T = [] ;
	return
end

if k(1)<=1 && k(end)>=size(x,1),		% crop includes all of the input data
	return
end
	
if ~isstruct(X),
	X = x(k,:) ;
	return
end
	
if length(fs)>1,
	X.data = [T x(k,:)] ;
else
	X.data = x(k,:) ;
end
X.crop = tcues ;
X.crop_units = 'seconds' ;
X.start_offset = tcues(1) ;
X.start_offset_units = 'seconds' ;

if ~isfield(X,'history') || isempty(X.history),
	X.history = 'crop_to' ;
else
	X.history = [X.history ',crop_to'] ;
end
